src/components/views/dialogs/AddressPickerDialog.js

```diff
@@ -383,31 +383,33 @@ module.exports = React.createClass({
     _processResults: function(results, query) {
         const queryList = [];
         results.forEach((result) => {
+            let newEntry;
+
             if (result.room_id) {
-                queryList.push({
+                newEntry = {
                     addressType: 'mx-room-id',
                     address: result.room_id,
                     displayName: result.name,
                     avatarMxc: result.avatar_url,
                     isKnown: true,
-                });
-                return;
-            }
-            if (!this.props.includeSelf &&
-                result.user_id === MatrixClientPeg.get().credentials.userId
-            ) {
-                return;
+                };
+            } else {
+                // Return objects, structure of which is defined
+                // by UserAddressType
+                newEntry = {
+                    addressType: 'mx-user-id',
+                    address: result.user_id,
+                    displayName: result.display_name,
+                    avatarMxc: result.avatar_url,
+                    isKnown: true,
+                };
             }
 
-            // Return objects, structure of which is defined
-            // by UserAddressType
-            queryList.push({
-                addressType: 'mx-user-id',
-                address: result.user_id,
-                displayName: result.display_name,
-                avatarMxc: result.avatar_url,
-                isKnown: true,
-            });
+            // Don't add it if it is in the excludedAddresses
+            if (this.props.excludedAddresses.find((entry) => {
```

Comment:
This is `O(n^2)` but optimising it would probably be premature.

Reply:
this one could be optimized by changing the format of exludedAddresses to be a map, for constant time access

Reply:
Yep, but I'd suggest we wait until there's an actual perf problem (so for now we choose readability over speed)